providers/implementations/signature/sm2sig.c

```diff
@@ -0,0 +1,502 @@
+    psm2ctx->flag_allow_md = 0;
+    return psm2ctx;
+}
+
+static int sm2sig_setup_md(PROV_SM2_CTX *ctx,
+                           const char *mdname, const char *mdprops)
+{
+    if (mdname != NULL) {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, mdname, mdprops);
+        int md_nid = sm2sig_get_md_nid(md);
+        size_t algorithmidentifier_len = 0;
+        const unsigned char *algorithmidentifier;
+
+        EVP_MD_free(ctx->md);
+        ctx->md = NULL;
+        ctx->mdname[0] = '\0';
+
+        algorithmidentifier =
+            sm2_algorithmidentifier_encoding(md_nid, &algorithmidentifier_len);
+
+        if (algorithmidentifier == NULL) {
+            EVP_MD_free(md);
+            return 0;
+        }
+
+        ctx->md = md;
+        OPENSSL_strlcpy(ctx->mdname, mdname, sizeof(ctx->mdname));
+        memcpy(ctx->aid, algorithmidentifier, algorithmidentifier_len);
+        ctx->aid_len = algorithmidentifier_len;
+    }
+    return 1;
+}
+
+static int sm2sig_signature_init(void *vpsm2ctx, void *ec)
+{
+    PROV_SM2_CTX *psm2ctx = (PROV_SM2_CTX *)vpsm2ctx;
+
+    if (psm2ctx == NULL || ec == NULL || !EC_KEY_up_ref(ec))
+        return 0;
+    EC_KEY_free(psm2ctx->ec);
+    psm2ctx->ec = ec;
+    return 1;
+}
+
+static int sm2sig_sign(void *vpsm2ctx, unsigned char *sig, size_t *siglen,
+                       size_t sigsize, const unsigned char *tbs, size_t tbslen)
+{
+    PROV_SM2_CTX *ctx = (PROV_SM2_CTX *)vpsm2ctx;
+    int ret;
+    unsigned int sltmp;
+    /* SM2 uses ECDSA_size as well */
+    size_t ecsize = ECDSA_size(ctx->ec);
+
+    if (sig == NULL) {
+        *siglen = ecsize;
+        return 1;
+    }
+
+    if (sigsize < (size_t)ecsize)
+        return 0;
+
+    if (ctx->mdsize != 0 && tbslen != ctx->mdsize)
+        return 0;
+
+    /* I would like to make the name something like SM2_sign... */
+    ret = sm2_sign(tbs, tbslen, sig, &sltmp, ctx->ec);
+    if (ret <= 0)
+        return 0;
+
+    *siglen = sltmp;
+    return 1;
+}
+
+static int sm2sig_verify(void *vpsm2ctx, const unsigned char *sig, size_t siglen,
+                         const unsigned char *tbs, size_t tbslen)
+{
+    PROV_SM2_CTX *ctx = (PROV_SM2_CTX *)vpsm2ctx;
+
+    if (ctx->mdsize != 0 && tbslen != ctx->mdsize)
+        return 0;
+
+    return sm2_verify(tbs, tbslen, sig, siglen, ctx->ec);
+}
+
+static void free_md(PROV_SM2_CTX *ctx)
+{
+    EVP_MD_CTX_free(ctx->mdctx);
+    EVP_MD_free(ctx->md);
+    ctx->mdctx = NULL;
+    ctx->md = NULL;
+    ctx->mdsize = 0;
+}
+
+static int sm2sig_digest_signverify_init(void *vpsm2ctx, const char *mdname,
+                                         const char *props, void *ec)
+{
+    PROV_SM2_CTX *ctx = (PROV_SM2_CTX *)vpsm2ctx;
+    size_t algorithmidentifier_len = 0;
+    const unsigned char *algorithmidentifier;
+
+    free_md(ctx);
+
+    if (!sm2sig_signature_init(vpsm2ctx, ec))
+        return 0;
+
+    ctx->md = EVP_MD_fetch(ctx->libctx, mdname, props);
```

Comment:
Should we simply disallow anything that isn't SM3?

Reply:
Not sure yet, all depend on how we treat https://github.com/openssl/openssl/pull/11248/files#r394931352